### ISSUE TITLE
Batch (parallelized) calculations of beta Shapley

### ISSUE THREAD

**Issue by kosmitive** [aai-institute/pyDVL#427]
Fine-grained marginal evaluations (used in our implementation of Beta Shapley) in combination with a parallel executor has sub-optimal performance. As a result it uses only a fraction of the computing power. I like the way Beta Shapley is implemented with the breakdown to marginal computations. To maintain that structure we should think about batching the marginal evaluations to increase performance. Obviously the optimal value depends on the used cores, but a batch size in between 30 and 50 can be chosen as a first guess.

**Comment by mdbenito** [comment]
Note that batch size will change the behaviour of algorithms because convergence criteria are checked in the main process. The most obvious case is maxchecks but all other stopping criteria will be evaluated only every batch_size samples.

The issue is with models that are trained very fast or sample sizes that are too small. I would have expected that using a process pool would alleviate the problem. Is joblib doing this?

Anyway, the proper solution is IMO not to batch but to use queues like we did a long time ago.

**Comment by kosmitive** [comment]
Yes but each sample of the batch is processed in order. I added a test case especially for checking that. The PR comes in handy for the current experiments to speed things up (by a factor of up to 14). We don't have to merge it in case we rely on a totally different pattern. 

My favorite here would be also task queues, although ProcessPoolExecutor probably has a task queue under the hood and as far as I am concerned the loky backend manges a ProcessPoolExecutor internally.

**Comment by mdbenito** [comment]
Come to think of it, process pools have running processes but they start afresh with each new task, and this can take too long. They cannot be using a task queue like I meant, since they couldn’t guarantee a clean slate if they did. We almost certainly want our own queues. 

As to your solution, given the massive speed up you mention, we can add it if we clearly document that the interface is temporary and will be deprecated. It is crucial to ensure that randomness is handled properly, though.

What do you think @AnesBenmerzoug ?

**Comment by AnesBenmerzoug** [comment]
@mdbenito I agree with what you said. Given the speed up that this brings I also think we should merge it.

We can go further in this direction by automatically finding the right batch number ([This](https://github.com/joblib/joblib/blob/bfd14ebc16ef36fde57b25654a10d5b21ca441d9/joblib/_parallel_backends.py#L304) is how it is done in joblib, for example) or we can consider once again using queues.

**Comment by mdbenito** [comment]
> We can go further in this direction by automatically finding the right batch number ([This](https://github.com/joblib/joblib/blob/bfd14ebc16ef36fde57b25654a10d5b21ca441d9/joblib/_parallel_backends.py#L304) is how it is done in joblib, for example) or we can consider once again using queues.

With the current idiom for job submission, timing would probably have to happen within the caller and not the sampler, since we take a bunch of samples in one go and submit the jobs.

**Comment by mdbenito** [comment]
@kosmitive I'm going to merge your PR (after some changes to documentation and deprecation notices that I've pushed), but I think batching belongs in the samplers. A simple way to do it is to wrap them in a generic @batched decorator like [this one I wrote](https://gist.github.com/mdbenito/7a964f3cdc2ec96380c75cb85e68618b)

**Comment by kosmitive** [comment]
> @kosmitive I'm going to merge your PR (after some changes to documentation and deprecation notices that I've pushed), but I think batching belongs in the samplers. A simple way to do it is to wrap them in a generic `batched` decorator like [this one I wrote](https://gist.github.com/mdbenito/7a964f3cdc2ec96380c75cb85e68618b) as a demo. I can add it some other time if we decide to keep the batching mechanism

Okay when I think about the sampler is indeed the clearer location. Also the decorator is way cleaner. But if we remove it latter it might be not the best to add it so deeply in the framework. So I leave the MR for editing and merging. 

@AnesBenmerzoug We could transfer it indirectly to the called processes, by introducing a stop queue, which also serves the speed recommendation of each worker.

**Comment by mdbenito** [comment]
> Okay when I think about the sampler is indeed the clearer location. Furthermore the decorator is way cleaner. But if we remove it latter it might be not the best to add it so deeply in the framework.

Good point. Although it only shifts the convention for samplers to return lists of samples instead of single items, which is not so bad. Whether or not using a decorator is good is not so clear. For instance we don't have explicit type information about the batched samplers and would have to create a protocol or maybe wrap the samplers at runtime, although yuk... Anyway, we can think about that some other time.
